File: weather-app/src/main/java/rcgonzalezf/org/weather/adapters/ModelAdapter.java
package rcgonzalezf.org.weather.adapters;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import androidx.databinding.BindingAdapter;
import androidx.recyclerview.widget.RecyclerView;
import org.rcgonzalezf.weather.common.models.WeatherViewModel;
import java.util.List;
import rcgonzalezf.org.weather.R;
import rcgonzalezf.org.weather.SettingsActivity;
import rcgonzalezf.org.weather.databinding.WeatherRowBinding;
import static rcgonzalezf.org.weather.utils.WeatherUtils.formatTemperature;

// TODO Revisit this class to do an upgrade to Kotlin
public class ModelAdapter<T extends WeatherViewModel>
        extends RecyclerView.Adapter<ModelAdapter.ModelViewHolder> implements View.OnClickListener {

    private List<T> models;
    private OnItemClickListener<WeatherViewModel> onItemClickListener;

    public ModelAdapter(List<T> models) {
        this.models = models;
    }

    @NonNull
    @Override
    public ModelViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        WeatherRowBinding weatherRowBinding =
                WeatherRowBinding.inflate(layoutInflater, parent, false);
        return new ModelViewHolder(weatherRowBinding);
    }

    @Override
    public void onBindViewHolder(ModelAdapter.ModelViewHolder holder, int position) {
        T mainModel = models.get(position);
        holder.bind(mainModel);
        populateTemperatureViews(holder.itemView.getContext(), holder, mainModel);
        holder.itemView.setOnClickListener(this);
        holder.itemView.setTag(mainModel);
    }

    @BindingAdapter("android:src")
    public static void setImageResource(ImageView imageView, int resource) {
        imageView.setImageResource(resource);
    }

    @VisibleForTesting
    void populateTemperatureViews(Context context,
            ModelAdapter.ModelViewHolder holder,
                                  T mainModel) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        boolean celsiusPreferred = prefs.getBoolean(SettingsActivity.PREF_TEMPERATURE_UNITS, true);

        final String celsiusText = formatTemperature(mainModel.getTemperature(), true, "C");
        final String fahrenheitText = formatTemperature(mainModel.getTemperature(), false, "F");

        holder.primaryTempTextView.setText(celsiusPreferred ? celsiusText : fahrenheitText);
        holder.secondaryTempTextView.setText(celsiusPreferred ? fahrenheitText : celsiusText);
    }

    @Override
    public int getItemCount() {
        return models.size();
    }

    public void setItems(List<T> models) {
        this.models = models;
        notifyDataSetChanged();
    }

    public void setOnItemClickListener(OnItemClickListener<WeatherViewModel> onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    @Override
    public void onClick(@NonNull final View view) {
        // Give some time to the ripple to finish the effect
        if (onItemClickListener != null) {
            new Handler().postDelayed(createClickRunnable(view), 200);
        }
    }

    @VisibleForTesting
    @NonNull
    Runnable createClickRunnable(@NonNull final View view) {
        return new Runnable() {
            @Override
            public void run() {
                onItemClickListener.onItemClick(view, (WeatherViewModel) view.getTag());
            }
        };
    }

    static class ModelViewHolder extends RecyclerView.ViewHolder {

        TextView secondaryTempTextView;
        TextView primaryTempTextView;
        final View itemView;
        private WeatherRowBinding weatherRowBinding;

        ModelViewHolder(WeatherRowBinding weatherRowBinding) {
            super(weatherRowBinding.getRoot());
            this.weatherRowBinding = weatherRowBinding;
            this.itemView = weatherRowBinding.getRoot();
            secondaryTempTextView =
                    (TextView) itemView.findViewById(R.id.secondary_temperature_text_view);
            primaryTempTextView = (TextView) itemView.findViewById(R.id.preferred_temperature_text_view);
        }

        public void bind(WeatherViewModel weatherViewModel) {
            weatherRowBinding.setWeather(weatherViewModel);
            weatherRowBinding.executePendingBindings();
        }
    }

    public interface OnItemClickListener<VM extends WeatherViewModel> {
        void onItemClick(View view, VM viewModel);
    }
}

File: weather-app/src/main/java/rcgonzalezf/org/weather/adapters/ModelAdapter.kt
package rcgonzalezf.org.weather.adapters

import android.content.Context
import android.os.Handler
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.annotation.VisibleForTesting
import androidx.databinding.BindingAdapter
import androidx.recyclerview.widget.RecyclerView
import org.rcgonzalezf.weather.common.models.WeatherViewModel
import rcgonzalezf.org.weather.SettingsActivity
import rcgonzalezf.org.weather.databinding.WeatherRowBinding
import rcgonzalezf.org.weather.utils.WeatherUtils.formatTemperature

class ModelAdapter<T : WeatherViewModel>(private var models: List<T>)
    : RecyclerView.Adapter<WeatherModelViewHolder>(), View.OnClickListener {
    private var onItemClickListener: OnItemClickListener<WeatherViewModel>? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WeatherModelViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val weatherRowBinding = WeatherRowBinding.inflate(layoutInflater, parent, false)
        return WeatherModelViewHolder(weatherRowBinding)
    }

    override fun onBindViewHolder(holder: WeatherModelViewHolder, position: Int) {
        val mainModel = models[position]
        holder.bind(mainModel)
        populateTemperatureViews(holder.itemViewRoot.context, holder, mainModel)
        holder.itemViewRoot.setOnClickListener(this)
        holder.itemViewRoot.tag = mainModel
    }

    @VisibleForTesting
    fun populateTemperatureViews(context: Context, holder: WeatherModelViewHolder,
                                 mainModel: T) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        val celsiusPreferred = prefs.getBoolean(SettingsActivity.PREF_TEMPERATURE_UNITS, true)
        val celsiusText = formatTemperature(mainModel.temperature, true, "C")
        val fahrenheitText = formatTemperature(mainModel.temperature, false, "F")
        holder.primaryTempTextView.text = if (celsiusPreferred) celsiusText else fahrenheitText
        holder.secondaryTempTextView.text = if (celsiusPreferred) fahrenheitText else celsiusText
    }

    override fun getItemCount(): Int {
        return models.size
    }

    fun setItems(models: List<T>) {
        this.models = models
        notifyDataSetChanged()
    }

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener<WeatherViewModel>?) {
        this.onItemClickListener = onItemClickListener
    }

    override fun onClick(view: View) {
        // Give some time to the ripple to finish the effect
        if (onItemClickListener != null) {
            Handler().postDelayed(createClickRunnable(view), 200)
        }
    }

    @VisibleForTesting
    fun createClickRunnable(view: View): Runnable {
        return Runnable { onItemClickListener?.onItemClick(view, view.tag as WeatherViewModel) }
    }

    interface OnItemClickListener<VM : WeatherViewModel> {
        fun onItemClick(view: View, viewModel: VM)
    }

    companion object {
        @JvmStatic
        @BindingAdapter("android:src")
        fun setImageResource(imageView: ImageView, resource: Int) {
            imageView.setImageResource(resource)
        }
    }
}
